zulip_bots: Add common tests to bot testing framework.

derAnfaenger opened the pull request:
This took some experimenting, since unittest doesn't cover this sort of thing. A nice effect is that we have a cheap jump for several bots to 100% coverage.

Because `test_lib.py` gets completely rewritten in the diff, I split out the first commit, which does not alter BotTestCase(TestCase) in any way. This should make it easier to review.

This addresses #122 and moves encrypt, followup, helloworld, help, incrementor, and weather to 100% coverage.

timabbott commented on zulip_bots/zulip_bots/test_lib.py:
```diff
@@ -26,130 +26,135 @@
-        # Strictly speaking, this function is not needed anymore,
-        # kept for now for legacy reasons.
-        self.call_request(message, expected_method, response)
+class BotTestCase:
```
I think it might be better to keep the existing code structure as `BotTestCaseBase`, and then have this double-layer class just inherit from that and contain a bit of code for the actual tests. That avoids indenting everything twice, and also provides an interface for a bot to inherit from BotTestCaseBase if it doesn't want the common usage tests.

Does that make sense?

derAnfaenger replied:
Yeah, that'd make it a lot nicer!

derAnfaenger commented:
Updated. Just moving the common tests to a new class `class BotTestCase(BotTestCaseBase)` actually didn't work, because unittest imports all module-level classes that inherit somehow from TestCase. So I added some code to our test runner `test-bots` that manually excludes all occurences of pure BotTestCase tests in the TestSuite.

timabbott commented:
Merged, with these changes for clarity and avoiding fragility:
```
diff --git a/tools/test-bots b/tools/test-bots
index 7db3c2e..c298bdb 100755
--- a/tools/test-bots
+++ b/tools/test-bots
@@ -98,7 +98,7 @@ def main():
         for test in tests:
             if isinstance(test, TestCase):
                 # Exclude test base class from being tested.
-                if 'BotTestCase' not in test.__class__.__name__:
+                if test.__class__.__name__ not in ['BotTestCase', 'BotTestCaseBase']:
                     filtered_tests.addTest(test)
             else:
                 filtered_tests.addTest(filter_tests(test))
diff --git a/zulip_bots/zulip_bots/test_lib.py b/zulip_bots/zulip_bots/test_lib.py
index f363cf7..48dd05d 100755
--- a/zulip_bots/zulip_bots/test_lib.py
+++ b/zulip_bots/zulip_bots/test_lib.py
@@ -27,6 +27,7 @@ from types import ModuleType
 from copy import deepcopy
 
 class BotTestCaseBase(TestCase):
+    """Test class for common Bot test helper methods"""
     bot_name = ''  # type: str
 
     def get_bot_message_handler(self):
@@ -155,6 +156,8 @@ class BotTestCaseBase(TestCase):
         self.call_request(message, expected_method, response)
 
 class BotTestCase(BotTestCaseBase):
+    """Test class extending BotTestCaseBase to add common default tests
+    that should be run (by default) for all our bots"""
     def test_bot_usage(self):
         # type: () -> None
         self.assertNotEqual(self.message_handler.usage(), '')
```
